// @openharness/core/dist/middleware/retry.js
import { isRetryableError, getRetryDelay, sleep } from "../utils.js";
const DEFAULT_RETRY = {
    maxRetries: 3,
    initialDelayMs: 1000,
    maxDelayMs: 30_000,
    backoffMultiplier: 2,
};
/**
 * Retries the inner runner on transient errors if no content has been yielded yet.
 * Yields `retry` lifecycle events on each retry attempt.
 */
export function withRetry(config) {
    const retryConfig = { ...DEFAULT_RETRY, ...config };
    return (runner) => async function* (history, input, options) {
        const { maxRetries } = retryConfig;
        const isRetryable = retryConfig.isRetryable ?? isRetryableError;
        for (let attempt = 0; attempt <= maxRetries; attempt++) {
            let hasYieldedContent = false;
            let shouldRetry = false;
            try {
                for await (const event of runner(history, input, options)) {
                    if (event.type === "text.delta" || event.type === "tool.start") {
                        hasYieldedContent = true;
                    }
                    if (event.type === "error" &&
                        !hasYieldedContent &&
                        isRetryable(event.error) &&
                        attempt < maxRetries) {
                        const delayMs = getRetryDelay(attempt, retryConfig, event.error);
                        yield {
                            type: "retry",
                            attempt,
                            maxRetries,
                            delayMs,
                            error: event.error,
                        };
                        await sleep(delayMs, options?.signal);
                        shouldRetry = true;
                        break;
                    }
                    if (event.type === "done") {
                        yield event;
                        break;
                    }
                    yield event;
                }
            }
            catch (thrown) {
                const error = thrown instanceof Error ? thrown : new Error(String(thrown));
                if (!hasYieldedContent && isRetryable(error) && attempt < maxRetries) {
                    const delayMs = getRetryDelay(attempt, retryConfig, error);
                    yield {
                        type: "retry",
                        attempt,
                        maxRetries,
                        delayMs,
                        error,
                    };
                    await sleep(delayMs, options?.signal);
                    continue;
                }
                throw thrown;
            }
            if (shouldRetry)
                continue;
            break;
        }
    };
}
