// @openharness/core/dist/agent-registry.js
const ALL_AWAIT_MODES = ["all", "any", "race", "allSettled"];
export function normalizeBackgroundConfig(input) {
    if (!input)
        return undefined;
    if (input === true) {
        return {
            maxConcurrent: Infinity,
            timeout: undefined,
            autoCancel: true,
            tools: { status: true, cancel: true, await: ALL_AWAIT_MODES },
        };
    }
    const awaitOpt = input.tools?.await;
    const awaitModes = awaitOpt === false
        ? false
        : awaitOpt === true || awaitOpt === undefined
            ? ALL_AWAIT_MODES
            : awaitOpt;
    return {
        maxConcurrent: input.maxConcurrent ?? Infinity,
        timeout: input.timeout,
        autoCancel: input.autoCancel ?? true,
        tools: {
            status: input.tools?.status ?? true,
            cancel: input.tools?.cancel ?? true,
            await: awaitModes,
        },
    };
}
// ── AgentRegistry ───────────────────────────────────────────────────
export class AgentRegistry {
    config;
    agents = new Map();
    idCounter = 0;
    constructor(config) {
        this.config = config;
    }
    get runningCount() {
        let count = 0;
        for (const entry of this.agents.values()) {
            if (entry.status === "running")
                count++;
        }
        return count;
    }
    /**
     * Spawn a child agent in the background. Returns an ID to track it.
     */
    spawn(name, child, prompt, opts) {
        if (this.config.maxConcurrent !== Infinity &&
            this.runningCount >= this.config.maxConcurrent) {
            throw new Error(`Maximum concurrent background agents (${this.config.maxConcurrent}) reached`);
        }
        const id = `bg-${++this.idCounter}`;
        const controller = new AbortController();
        // Link to parent abort signal
        if (opts.signal) {
            opts.signal.addEventListener("abort", () => controller.abort(), { once: true });
        }
        const entry = {
            promise: undefined,
            controller,
            status: "running",
            sessionId: opts.sessionId,
        };
        // Auto-timeout
        if (this.config.timeout) {
            entry.timeoutId = setTimeout(() => {
                if (entry.status === "running") {
                    controller.abort();
                    entry.status = "cancelled";
                    entry.error = `Timed out after ${this.config.timeout}ms`;
                }
            }, this.config.timeout);
        }
        entry.promise = (async () => {
            try {
                let lastText = "";
                for await (const event of child.run([], prompt, { signal: controller.signal })) {
                    opts.onEvent?.([name], event);
                    if (event.type === "text.done") {
                        lastText = event.text;
                    }
                }
                entry.status = "done";
                entry.result = lastText || "(no output)";
                return entry.result;
            }
            catch (e) {
                if (controller.signal.aborted && entry.status !== "cancelled") {
                    entry.status = "cancelled";
                }
                else if (entry.status === "running") {
                    entry.status = "failed";
                }
                entry.error = e instanceof Error ? e.message : String(e);
                throw e;
            }
            finally {
                if (entry.timeoutId)
                    clearTimeout(entry.timeoutId);
                await child.close();
            }
        })();
        // Prevent unhandled rejection warnings — errors are tracked in entry.status/error
        entry.promise.catch(() => { });
        this.agents.set(id, entry);
        return id;
    }
    /**
     * Non-blocking status check.
     */
    getStatus(id) {
        const entry = this.agents.get(id);
        if (!entry)
            return undefined;
        return {
            status: entry.status,
            sessionId: entry.sessionId,
            result: entry.result,
            error: entry.error,
        };
    }
    /**
     * Cancel a running background agent.
     */
    cancel(id) {
        const entry = this.agents.get(id);
        if (!entry || entry.status !== "running")
            return false;
        entry.controller.abort();
        entry.status = "cancelled";
        if (entry.timeoutId)
            clearTimeout(entry.timeoutId);
        return true;
    }
    /**
     * Promise.all — wait for all to succeed. Rejects on first failure.
     */
    async awaitAll(ids) {
        const results = new Map();
        await Promise.all(ids.map(async (id) => {
            const entry = this.agents.get(id);
            if (!entry)
                throw new Error(`Agent "${id}" not found`);
            try {
                const result = await entry.promise;
                results.set(id, result);
            }
            catch {
                throw new Error(`Agent "${id}" failed: ${entry.error ?? "unknown error"}`);
            }
        }));
        return results;
    }
    /**
     * Promise.allSettled — wait for all, return results and errors.
     */
    async awaitAllSettled(ids) {
        const results = new Map();
        await Promise.allSettled(ids.map(async (id) => {
            const entry = this.agents.get(id);
            if (!entry) {
                results.set(id, { status: "failed", error: `Agent "${id}" not found` });
                return;
            }
            try {
                const result = await entry.promise;
                results.set(id, {
                    status: "done",
                    sessionId: entry.sessionId,
                    result,
                });
            }
            catch {
                results.set(id, {
                    status: entry.status === "cancelled" ? "cancelled" : "failed",
                    sessionId: entry.sessionId,
                    error: entry.error,
                });
            }
        }));
        return results;
    }
    /**
     * Promise.any — first success wins. Rejects only if all fail.
     */
    async awaitAny(ids) {
        return Promise.any(ids.map(async (id) => {
            const entry = this.agents.get(id);
            if (!entry)
                throw new Error(`Agent "${id}" not found`);
            const result = await entry.promise;
            return { id, sessionId: entry.sessionId, result };
        }));
    }
    /**
     * Promise.race — first to settle (success or failure) wins.
     */
    async awaitRace(ids) {
        return Promise.race(ids.map(async (id) => {
            const entry = this.agents.get(id);
            if (!entry)
                return { id, error: `Agent "${id}" not found` };
            try {
                const result = await entry.promise;
                return { id, sessionId: entry.sessionId, result };
            }
            catch {
                return {
                    id,
                    sessionId: entry.sessionId,
                    error: entry.error ?? "unknown error",
                };
            }
        }));
    }
    /**
     * Cancel all running background agents.
     */
    cancelAll() {
        for (const entry of this.agents.values()) {
            if (entry.status === "running") {
                entry.controller.abort();
                entry.status = "cancelled";
                if (entry.timeoutId)
                    clearTimeout(entry.timeoutId);
            }
        }
    }
}
