// @openharness/core/dist/agent.js
import { tool, streamText, stepCountIs, } from "ai";
import { z } from "zod";
import { Session } from "./session.js";
import { loadInstructions } from "./instructions.js";
import { connectMCPServers, closeMCPClients, } from "./mcp.js";
import { discoverSkills, } from "./skills.js";
import { InMemorySubagentSessionMetadataStore, isSubagentCatalog, } from "./subagents.js";
import { createSkillTool } from "./tools/skill.js";
import { AgentRegistry, normalizeBackgroundConfig, } from "./agent-registry.js";
const SUBAGENT_SESSION_RUNTIME = new WeakMap();
const TASK_SESSION_MODES = ["stateless", "new", "resume", "fork"];
// ── Agent ────────────────────────────────────────────────────────────
export class Agent {
    name;
    description;
    model;
    systemPrompt;
    maxSteps;
    temperature;
    maxTokens;
    instructions;
    maxSubagentDepth;
    approve;
    onSubagentEvent;
    /** Original subagent templates or catalog (stored for nested children). */
    subagents;
    /** Optional resumable subagent session config. */
    subagentSessions;
    /** Background config (stored so nested children can inherit it). */
    subagentBackground;
    /** Registry for background subagents. Only present when configured. */
    agentRegistry;
    /** Static tools provided at construction time. */
    tools;
    /** MCP server configs — connected lazily on first run. */
    mcpServerConfigs;
    mcpConnection = null;
    /** Skills config — discovered lazily on first run. */
    skillsConfig;
    cachedSkills = null;
    cachedInstructions = null;
    constructor(options) {
        this.name = options.name;
        this.description = options.description;
        this.model = options.model;
        this.systemPrompt = options.systemPrompt;
        this.maxSteps = options.maxSteps ?? 100;
        this.temperature = options.temperature;
        this.maxTokens = options.maxTokens;
        this.instructions = options.instructions ?? true;
        this.maxSubagentDepth = options.maxSubagentDepth ?? 1;
        this.approve = options.approve;
        this.onSubagentEvent = options.onSubagentEvent;
        this.subagents = options.subagents;
        this.subagentSessions = options.subagentSessions;
        this.subagentBackground = options.subagentBackground;
        this.mcpServerConfigs = options.mcpServers;
        this.skillsConfig = options.skills;
        this.tools = options.tools;
        if (options.subagentSessions) {
            getSubagentSessionRuntime(options.subagentSessions);
        }
        if (options.subagents && this.maxSubagentDepth > 0 && options.subagentBackground) {
            const bgConfig = normalizeBackgroundConfig(options.subagentBackground);
            if (bgConfig) {
                this.agentRegistry = new AgentRegistry(bgConfig);
            }
        }
    }
    /**
     * Close all MCP server connections. Call this when the agent is no longer needed.
     */
    async close() {
        if (this.agentRegistry) {
            this.agentRegistry.cancelAll();
        }
        if (this.mcpConnection) {
            await closeMCPClients(this.mcpConnection.clients);
            this.mcpConnection = null;
        }
    }
    async *run(history, input, options) {
        const messages = [...history];
        if (typeof input === "string") {
            messages.push({ role: "user", content: input });
        }
        else {
            messages.push(...input);
        }
        if (this.instructions && this.cachedInstructions === null) {
            this.cachedInstructions = await loadInstructions();
        }
        if (this.mcpServerConfigs && !this.mcpConnection) {
            this.mcpConnection = await connectMCPServers(this.mcpServerConfigs);
        }
        if (this.skillsConfig && this.cachedSkills === null) {
            this.cachedSkills = await discoverSkills(this.skillsConfig);
        }
        const systemParts = [this.systemPrompt, this.cachedInstructions].filter(Boolean);
        const system = systemParts.length > 0 ? systemParts.join("\n\n") : undefined;
        const subagentTools = await createSubagentTools(this.subagents, this.maxSubagentDepth, this.onSubagentEvent, this.agentRegistry, this.subagentBackground, this.subagentSessions);
        const allTools = {
            ...(this.cachedSkills?.length
                ? { skill: createSkillTool(this.cachedSkills) }
                : {}),
            ...(this.mcpConnection?.tools ?? {}),
            ...(this.tools ?? {}),
            ...(subagentTools ?? {}),
        };
        const tools = this.approve && Object.keys(allTools).length > 0
            ? wrapToolsWithApproval(allTools, this.approve)
            : Object.keys(allTools).length > 0
                ? allTools
                : undefined;
        const stream = streamText({
            model: this.model,
            system,
            messages,
            tools,
            stopWhen: stepCountIs(this.maxSteps),
            temperature: this.temperature,
            maxOutputTokens: this.maxTokens,
            abortSignal: options?.signal,
        });
        let stepNumber = 0;
        let stepText = "";
        let stepReasoning = "";
        const completedStepMessages = [];
        let doneEmitted = false;
        let abortReason;
        try {
            for await (const part of stream.fullStream) {
                switch (part.type) {
                    case "start-step":
                        stepNumber++;
                        stepText = "";
                        stepReasoning = "";
                        yield { type: "step.start", stepNumber };
                        break;
                    case "text-delta":
                        stepText += part.text;
                        yield { type: "text.delta", text: part.text };
                        break;
                    case "text-end":
                        if (stepText) {
                            yield { type: "text.done", text: stepText };
                        }
                        break;
                    case "reasoning-delta":
                        stepReasoning += part.text;
                        yield { type: "reasoning.delta", text: part.text };
                        break;
                    case "reasoning-end":
                        if (stepReasoning) {
                            yield { type: "reasoning.done", text: stepReasoning };
                        }
                        break;
                    case "tool-call":
                        yield {
                            type: "tool.start",
                            toolCallId: part.toolCallId,
                            toolName: part.toolName,
                            input: part.input,
                        };
                        break;
                    case "tool-result":
                        yield {
                            type: "tool.done",
                            toolCallId: part.toolCallId,
                            toolName: part.toolName,
                            output: part.output,
                        };
                        break;
                    case "tool-error":
                        yield {
                            type: "tool.error",
                            toolCallId: part.toolCallId,
                            toolName: part.toolName,
                            error: String(part.error),
                        };
                        break;
                    case "finish-step":
                        completedStepMessages.push(...buildAbortedStepMessages(stepText, stepReasoning));
                        stepText = "";
                        stepReasoning = "";
                        yield {
                            type: "step.done",
                            stepNumber,
                            usage: toTokenUsage(part.usage),
                            finishReason: part.finishReason,
                        };
                        break;
                    case "error":
                        yield {
                            type: "error",
                            error: part.error instanceof Error
                                ? part.error
                                : new Error(String(part.error)),
                        };
                        break;
                    case "finish": {
                        const result = part.finishReason === "stop"
                            ? "complete"
                            : part.finishReason === "tool-calls"
                                ? "max_steps"
                                : part.finishReason === "error"
                                    ? "error"
                                    : "stopped";
                        const response = await stream.response;
                        messages.push(...response.messages);
                        doneEmitted = true;
                        yield {
                            type: "done",
                            result,
                            messages,
                            totalUsage: toTokenUsage(part.totalUsage),
                        };
                        break;
                    }
                    case "abort":
                        abortReason = part.reason ?? "aborted";
                        break;
                }
            }
            if (!doneEmitted && abortReason !== undefined) {
                yield {
                    type: "error",
                    error: new Error(abortReason),
                };
                yield {
                    type: "done",
                    result: "stopped",
                    messages: [
                        ...messages,
                        ...completedStepMessages,
                        ...buildAbortedStepMessages(stepText, stepReasoning),
                    ],
                    totalUsage: {
                        inputTokens: undefined,
                        outputTokens: undefined,
                        totalTokens: undefined,
                    },
                };
            }
        }
        catch (error) {
            yield {
                type: "error",
                error: error instanceof Error ? error : new Error(String(error)),
            };
            yield {
                type: "done",
                result: "error",
                messages,
                totalUsage: {
                    inputTokens: undefined,
                    outputTokens: undefined,
                    totalTokens: undefined,
                },
            };
        }
    }
}
// ── Subagent tools ──────────────────────────────────────────────────
async function createSubagentTools(subagents, remainingDepth, onSubagentEvent, registry, backgroundConfig, sessionConfig) {
    if (!subagents || remainingDepth <= 0)
        return undefined;
    if (Array.isArray(subagents) && subagents.length === 0)
        return undefined;
    const bgConfig = registry && backgroundConfig
        ? normalizeBackgroundConfig(backgroundConfig)
        : undefined;
    const task = await createTaskTool(subagents, remainingDepth, onSubagentEvent, registry, sessionConfig);
    return {
        task,
        ...(registry && bgConfig?.tools.status
            ? { agent_status: createStatusTool(registry) }
            : {}),
        ...(registry && bgConfig?.tools.cancel
            ? { agent_cancel: createCancelTool(registry) }
            : {}),
        ...(registry && bgConfig?.tools.await
            ? {
                agent_await: createAwaitTool(registry, bgConfig.tools.await),
            }
            : {}),
    };
}
function createChildFromTemplate(template, remainingDepth, onSubagentEvent) {
    const nextDepth = remainingDepth - 1;
    const childOnSubagentEvent = onSubagentEvent
        ? (childPath, event) => onSubagentEvent([template.name, ...childPath], event)
        : undefined;
    return new Agent({
        name: template.name,
        description: template.description,
        model: template.model,
        systemPrompt: template.systemPrompt,
        tools: template.tools,
        maxSteps: template.maxSteps,
        temperature: template.temperature,
        maxTokens: template.maxTokens,
        instructions: template.instructions,
        ...(nextDepth > 0 && template.subagents
            ? {
                subagents: template.subagents,
                subagentSessions: template.subagentSessions,
                maxSubagentDepth: nextDepth,
                onSubagentEvent: childOnSubagentEvent,
                subagentBackground: template.subagentBackground,
            }
            : {}),
    });
}
async function createTaskTool(subagents, remainingDepth, onSubagentEvent, registry, sessionConfig) {
    const descriptors = await listSubagentDescriptors(subagents);
    const names = descriptors.map((subagent) => subagent.name);
    const listing = descriptors
        .map((subagent) => `- ${subagent.name}: ${subagent.description ?? subagent.name}`)
        .join("\n");
    const descriptionLines = [
        "Spawn a subagent to handle a task autonomously.",
        "The subagent runs with its own tools, completes the work, and returns the result.",
        "Launch multiple agents concurrently when possible by calling this tool multiple times in one response.",
    ];
    if (sessionConfig) {
        descriptionLines.push("", "Set session.mode=new to create a resumable subagent session.", "Set session.mode=resume with session.id to continue an earlier subagent session.", "Set session.mode=fork with session.id to clone an earlier session into a new one.", 'When session is omitted, the default mode is "' +
            (sessionConfig.defaultMode ?? "stateless") +
            '".');
    }
    if (registry) {
        descriptionLines.push("", "Set background=true to spawn the agent in the background and return immediately with a run ID.", "Use agent_status, agent_await, or agent_cancel to manage background runs.");
    }
    if (listing) {
        descriptionLines.push("", "Available agents:", listing);
    }
    else if (isSubagentCatalog(subagents)) {
        descriptionLines.push("", "Available agents are resolved dynamically at runtime.");
    }
    const baseSchema = z.object({
        agent: createAgentSelectionSchema(names).describe("Which agent to use"),
        prompt: z.string().describe("Detailed task description for the subagent"),
    });
    const withSession = sessionConfig
        ? baseSchema.extend({
            session: createTaskSessionSchema()
                .optional()
                .describe("Optional resumable session instructions"),
        })
        : baseSchema;
    const inputSchema = registry
        ? withSession.extend({
            background: z
                .boolean()
                .optional()
                .describe("If true, spawn in background and return immediately with a run ID"),
        })
        : withSession;
    return tool({
        description: descriptionLines.join("\n"),
        inputSchema,
        execute: async (rawInput, { abortSignal }) => {
            const { agent: agentName, prompt } = rawInput;
            const background = "background" in rawInput ? rawInput.background : undefined;
            const session = "session" in rawInput ? rawInput.session : undefined;
            const template = await resolveSubagent(subagents, agentName);
            if (!template) {
                const suffix = names.length > 0 ? ` Available agents: ${names.join(", ")}.` : "";
                throw new Error(`Unknown subagent "${agentName}".${suffix}`);
            }
            const prepared = sessionConfig
                ? await prepareSubagentChild({
                    agentName,
                    template,
                    remainingDepth,
                    onSubagentEvent,
                    sessionConfig,
                    session,
                })
                : {
                    child: createChildFromTemplate(template, remainingDepth, onSubagentEvent),
                    sessionId: undefined,
                };
            const { child, sessionId } = prepared;
            if (background && registry) {
                const id = registry.spawn(agentName, child, prompt, {
                    signal: abortSignal,
                    onEvent: onSubagentEvent,
                    sessionId,
                });
                return formatBackgroundSpawn(agentName, id, sessionId);
            }
            let lastText = "";
            try {
                for await (const event of child.run([], prompt, { signal: abortSignal })) {
                    onSubagentEvent?.([agentName], event);
                    if (event.type === "text.done") {
                        lastText = event.text;
                    }
                }
            }
            finally {
                await child.close();
            }
            return formatTaskResult(lastText || "(no output)", sessionId);
        },
    });
}
function createStatusTool(registry) {
    return tool({
        description: "Check the status of a background run without blocking.",
        inputSchema: z.object({
            id: z.string().describe("The run ID returned by a background task spawn"),
        }),
        execute: async ({ id }) => {
            const status = registry.getStatus(id);
            if (!status)
                return `Agent "${id}" not found.`;
            if (status.status === "done") {
                return formatAgentStatus(id, status.status, status.result, undefined, status.sessionId);
            }
            if (status.status === "failed") {
                return formatAgentStatus(id, status.status, undefined, status.error ?? "unknown", status.sessionId);
            }
            if (status.status === "cancelled") {
                return formatAgentStatus(id, status.status, undefined, undefined, status.sessionId);
            }
            return formatAgentStatus(id, status.status, undefined, undefined, status.sessionId);
        },
    });
}
function createCancelTool(registry) {
    return tool({
        description: "Cancel a running background run.",
        inputSchema: z.object({
            id: z.string().describe("The run ID to cancel"),
        }),
        execute: async ({ id }) => {
            const cancelled = registry.cancel(id);
            return cancelled
                ? `Agent "${id}" cancelled.`
                : `Agent "${id}" is not running (may have already completed or been cancelled).`;
        },
    });
}
function createAwaitTool(registry, modes) {
    const modeDescriptions = {
        all: '"all": Wait for all agents to succeed. Fails fast if any agent fails.',
        allSettled: '"allSettled": Wait for all agents to finish. Returns both results and errors.',
        any: '"any": Wait for the first agent to succeed. Only fails if all agents fail.',
        race: '"race": Wait for the first agent to settle (succeed or fail).',
    };
    return tool({
        description: [
            "Wait for one or more background runs to complete.",
            "",
            "Modes:",
            ...modes.map((mode) => `- ${modeDescriptions[mode]}`),
        ].join("\n"),
        inputSchema: z.object({
            ids: z.array(z.string()).min(1).describe("Run IDs to wait for"),
            mode: z.enum(modes).describe("How to wait for runs"),
        }),
        execute: async ({ ids, mode }) => {
            switch (mode) {
                case "all": {
                    try {
                        const results = await registry.awaitAll(ids);
                        const entries = [...results.entries()].map(([id, result]) => {
                            const sessionId = registry.getStatus(id)?.sessionId;
                            return `<agent id="${id}"${formatOptionalAttr("session_id", sessionId)}>\n${result}\n</agent>`;
                        });
                        return `<await_result mode="all">\n${entries.join("\n")}\n</await_result>`;
                    }
                    catch (error) {
                        const message = error instanceof Error ? error.message : String(error);
                        return `<await_result mode="all" error="${message}" />`;
                    }
                }
                case "allSettled": {
                    const results = await registry.awaitAllSettled(ids);
                    const entries = [...results.entries()].map(([id, result]) => {
                        const sessionAttr = formatOptionalAttr("session_id", result.sessionId);
                        if (result.status === "done") {
                            return `<agent id="${id}" status="done"${sessionAttr}>\n${result.result}\n</agent>`;
                        }
                        return `<agent id="${id}" status="${result.status}"${sessionAttr} error="${result.error ?? "unknown"}" />`;
                    });
                    return `<await_result mode="allSettled">\n${entries.join("\n")}\n</await_result>`;
                }
                case "any": {
                    try {
                        const { id, sessionId, result } = await registry.awaitAny(ids);
                        return `<await_result mode="any" winner="${id}"${formatOptionalAttr("session_id", sessionId)}>\n${result}\n</await_result>`;
                    }
                    catch {
                        return `<await_result mode="any" error="All agents failed." />`;
                    }
                }
                case "race": {
                    const settled = await registry.awaitRace(ids);
                    const sessionAttr = formatOptionalAttr("session_id", settled.sessionId);
                    if (settled.error) {
                        return `<await_result mode="race" settled="${settled.id}"${sessionAttr} status="failed" error="${settled.error}" />`;
                    }
                    return `<await_result mode="race" settled="${settled.id}"${sessionAttr}>\n${settled.result}\n</await_result>`;
                }
            }
        },
    });
}
async function prepareSubagentChild(params) {
    const { agentName, template, remainingDepth, onSubagentEvent, sessionConfig, session, } = params;
    const child = createChildFromTemplate(template, remainingDepth, onSubagentEvent);
    const mode = session?.mode ?? sessionConfig.defaultMode ?? "stateless";
    if (mode === "stateless") {
        return { child };
    }
    const runtime = getSubagentSessionRuntime(sessionConfig);
    const now = new Date().toISOString();
    let sessionId;
    let initialMessages = [];
    let metadata;
    try {
        switch (mode) {
            case "new":
                sessionId = crypto.randomUUID();
                metadata = {
                    sessionId,
                    agentName,
                    createdAt: now,
                    updatedAt: now,
                };
                await runtime.metadataStore.save(metadata);
                break;
            case "resume": {
                sessionId = requireSessionId(session, mode);
                metadata = await loadRequiredSessionMetadata(runtime.metadataStore, sessionId, agentName);
                const storedMessages = await sessionConfig.messages.load(sessionId);
                if (!storedMessages) {
                    throw new Error(`Subagent session "${sessionId}" could not be loaded.`);
                }
                initialMessages = structuredClone(storedMessages);
                metadata = { ...metadata, updatedAt: now };
                break;
            }
            case "fork": {
                const sourceId = requireSessionId(session, mode);
                await loadRequiredSessionMetadata(runtime.metadataStore, sourceId, agentName);
                const sourceMessages = await sessionConfig.messages.load(sourceId);
                if (!sourceMessages) {
                    throw new Error(`Subagent session "${sourceId}" could not be loaded.`);
                }
                sessionId = crypto.randomUUID();
                initialMessages = structuredClone(sourceMessages);
                metadata = {
                    sessionId,
                    agentName,
                    createdAt: now,
                    updatedAt: now,
                };
                await runtime.metadataStore.save(metadata);
                await sessionConfig.messages.save(sessionId, initialMessages);
                break;
            }
            default:
                return { child };
        }
    }
    catch (error) {
        await child.close();
        throw error;
    }
    if (runtime.activeSessionIds.has(sessionId)) {
        await child.close();
        throw new Error(`Subagent session "${sessionId}" is already running.`);
    }
    runtime.activeSessionIds.add(sessionId);
    let released = false;
    const close = async () => {
        if (released)
            return;
        released = true;
        runtime.activeSessionIds.delete(sessionId);
        await child.close();
    };
    const statefulChild = {
        run: async function* (_history, input, options) {
            const session = new Session({
                agent: child,
                sessionId,
                sessionStore: sessionConfig.messages,
                ...(sessionConfig.sessionOptions ?? {}),
            });
            session.messages = structuredClone(initialMessages);
            try {
                for await (const event of session.send(input, options)) {
                    if (event.type === "turn.start" ||
                        event.type === "turn.done" ||
                        event.type === "compaction.start" ||
                        event.type === "compaction.pruned" ||
                        event.type === "compaction.summary" ||
                        event.type === "compaction.done" ||
                        event.type === "retry") {
                        continue;
                    }
                    yield event;
                }
                await runtime.metadataStore.save({
                    sessionId,
                    agentName,
                    createdAt: metadata.createdAt,
                    updatedAt: new Date().toISOString(),
                });
            }
            catch (error) {
                await runtime.metadataStore.save({
                    sessionId,
                    agentName,
                    createdAt: metadata.createdAt,
                    updatedAt: new Date().toISOString(),
                });
                throw error;
            }
        },
        close,
    };
    return { child: statefulChild, sessionId };
}
function getSubagentSessionRuntime(config) {
    let runtime = SUBAGENT_SESSION_RUNTIME.get(config);
    if (!runtime) {
        runtime = {
            activeSessionIds: new Set(),
            metadataStore: config.metadata ?? new InMemorySubagentSessionMetadataStore(),
        };
        SUBAGENT_SESSION_RUNTIME.set(config, runtime);
    }
    return runtime;
}
function createAgentSelectionSchema(names) {
    if (names.length > 0) {
        return z.enum(names);
    }
    return z.string().min(1);
}
function createTaskSessionSchema() {
    return z
        .object({
        mode: z.enum(TASK_SESSION_MODES).describe("How to handle subagent memory"),
        id: z.string().optional().describe("Existing session ID to resume or fork"),
    })
        .superRefine((value, ctx) => {
        if ((value.mode === "resume" || value.mode === "fork") && !value.id) {
            ctx.addIssue({
                code: z.ZodIssueCode.custom,
                message: `session.id is required when session.mode="${value.mode}"`,
                path: ["id"],
            });
        }
        if ((value.mode === "stateless" || value.mode === "new") && value.id) {
            ctx.addIssue({
                code: z.ZodIssueCode.custom,
                message: `session.id is not used when session.mode="${value.mode}"`,
                path: ["id"],
            });
        }
    });
}
async function listSubagentDescriptors(subagents) {
    if (Array.isArray(subagents)) {
        return subagents.map((agent) => ({
            name: agent.name,
            description: agent.description,
        }));
    }
    return subagents.list();
}
async function resolveSubagent(subagents, name) {
    if (Array.isArray(subagents)) {
        return subagents.find((agent) => agent.name === name);
    }
    return subagents.resolve(name);
}
async function loadRequiredSessionMetadata(store, sessionId, agentName) {
    const metadata = await store.load(sessionId);
    if (!metadata) {
        throw new Error(`Unknown subagent session "${sessionId}".`);
    }
    if (metadata.agentName !== agentName) {
        throw new Error(`Subagent session "${sessionId}" belongs to "${metadata.agentName}", not "${agentName}".`);
    }
    return metadata;
}
function requireSessionId(session, mode) {
    if (!session?.id) {
        throw new Error(`session.id is required when session.mode="${mode}".`);
    }
    return session.id;
}
function formatBackgroundSpawn(agentName, runId, sessionId) {
    return `<background_spawn agent_id="${runId}"${formatOptionalAttr("session_id", sessionId)}>\nAgent "${agentName}" spawned in background with id "${runId}".\nUse agent_status, agent_await, or agent_cancel to manage it.\n</background_spawn>`;
}
function formatTaskResult(result, sessionId) {
    return `<task_result${formatOptionalAttr("session_id", sessionId)}>\n${result}\n</task_result>`;
}
function formatAgentStatus(runId, status, result, error, sessionId) {
    const sessionAttr = formatOptionalAttr("session_id", sessionId);
    if (status === "done") {
        return `<agent_status id="${runId}" status="done"${sessionAttr}>\n${result}\n</agent_status>`;
    }
    if (status === "failed") {
        return `<agent_status id="${runId}" status="failed"${sessionAttr} error="${error ?? "unknown"}" />`;
    }
    if (status === "cancelled") {
        return `<agent_status id="${runId}" status="cancelled"${sessionAttr} />`;
    }
    return `<agent_status id="${runId}" status="running"${sessionAttr} />`;
}
function formatOptionalAttr(name, value) {
    return value ? ` ${name}="${value}"` : "";
}
// ── Helpers ──────────────────────────────────────────────────────────
function toTokenUsage(usage) {
    return {
        inputTokens: usage.inputTokens,
        outputTokens: usage.outputTokens,
        totalTokens: usage.totalTokens,
    };
}
function buildAbortedStepMessages(text, reasoning) {
    const parts = [];
    if (reasoning) {
        parts.push({ type: "reasoning", text: reasoning });
    }
    if (text) {
        parts.push({ type: "text", text });
    }
    if (parts.length === 0)
        return [];
    return [
        {
            role: "assistant",
            content: parts.length === 1 && parts[0].type === "text" ? text : parts,
        },
    ];
}
function wrapToolsWithApproval(tools, approve) {
    const wrapped = {};
    for (const [name, toolDef] of Object.entries(tools)) {
        if (!toolDef.execute) {
            wrapped[name] = toolDef;
            continue;
        }
        const originalExecute = toolDef.execute;
        wrapped[name] = {
            ...toolDef,
            execute: async (input, options) => {
                const allowed = await approve({
                    toolName: name,
                    toolCallId: options.toolCallId,
                    input,
                });
                if (!allowed) {
                    throw new ToolDeniedError(name);
                }
                return originalExecute(input, options);
            },
        };
    }
    return wrapped;
}
export class ToolDeniedError extends Error {
    constructor(toolName) {
        super(`Tool call to "${toolName}" was denied.`);
        this.name = "ToolDeniedError";
    }
}
