// @openharness/core/dist/conversation.js
import { createUIMessageStreamResponse } from "ai";
import { sessionEventsToUIStream } from "./ui-stream.js";
/**
 * Thin stateful wrapper over a composed Runner.
 * Manages `messages` — updates from `done` events.
 * No compaction/retry/hooks/turn logic — those are middleware concerns.
 */
export class Conversation {
    messages = [];
    sessionId;
    runner;
    store;
    constructor(options) {
        this.runner = options.runner;
        this.sessionId = options.sessionId ?? crypto.randomUUID();
        this.store = options.store;
    }
    /**
     * Send a message through the composed runner pipeline.
     * Updates `this.messages` from `done` events automatically.
     */
    async *send(input, options) {
        for await (const event of this.runner(this.messages, input, options)) {
            if (event.type === "done") {
                this.messages = event.messages;
            }
            yield event;
        }
    }
    /** Load messages from store. */
    async load() {
        if (!this.store)
            return false;
        const loaded = await this.store.load(this.sessionId);
        if (loaded) {
            this.messages = loaded;
            return true;
        }
        return false;
    }
    /** Save messages to store. */
    async save() {
        if (!this.store)
            return;
        await this.store.save(this.sessionId, this.messages);
    }
    /**
     * Convert a send call into an AI SDK 5 UIMessage stream.
     * Reuses the existing sessionEventsToUIStream function.
     */
    toUIMessageStream(input, options) {
        return sessionEventsToUIStream(this.send(input, options), options);
    }
    /**
     * Convert a send call into an HTTP Response with SSE encoding.
     * Ready to return from any HTTP handler (Next.js, Express, etc.).
     */
    toResponse(input, init) {
        const { signal, ...responseInit } = init ?? {};
        return createUIMessageStreamResponse({
            stream: this.toUIMessageStream(input, { signal }),
            ...responseInit,
        });
    }
}
