// @openharness/core/dist/session.js
import { generateText, createUIMessageStreamResponse, } from "ai";
import { sessionEventsToUIStream } from "./ui-stream.js";
import { isRetryableError as _isRetryableError, getRetryDelay as _getRetryDelay, sleep as _sleep, addUsage as _addUsage, defaultEstimateTokens as _defaultEstimateTokens, } from "./utils.js";
const DEFAULT_RETRY = {
    maxRetries: 3,
    initialDelayMs: 1000,
    maxDelayMs: 30_000,
    backoffMultiplier: 2,
};
// ── Default Compaction Strategy ─────────────────────────────────────
const DEFAULT_SUMMARY_PROMPT = `Summarize this conversation for the next agent turn:
1. Goal: What the user is trying to accomplish
2. Instructions: Key directives and constraints mentioned
3. Discoveries: Important findings during the conversation
4. Accomplished: What's been completed, files changed, actions taken
5. Current State: Where things stand, pending work
6. Relevant Context: File paths, code snippets, specific details needed to continue`;
export class DefaultCompactionStrategy {
    protectedTokens;
    minPruneSavings;
    summaryModel;
    summaryPrompt;
    estimateTokens;
    constructor(options) {
        this.protectedTokens = options?.protectedTokens ?? 40_000;
        this.minPruneSavings = options?.minPruneSavings ?? 20_000;
        this.summaryModel = options?.summaryModel;
        this.summaryPrompt = options?.summaryPrompt;
        this.estimateTokens = options?.estimateTokens ?? defaultEstimateTokens;
    }
    async compact(context) {
        // Phase 1: Pruning
        const pruned = pruneToolResults(context.messages, this.protectedTokens, this.minPruneSavings, this.estimateTokens);
        if (pruned.tokensSaved >= this.minPruneSavings) {
            return {
                messages: pruned.messages,
                messagesRemoved: 0,
                tokensPruned: pruned.tokensSaved,
            };
        }
        // Phase 2: Summarization
        const model = this.summaryModel ?? context.model;
        const prompt = context.compactionPrompt ?? this.summaryPrompt ?? DEFAULT_SUMMARY_PROMPT;
        const conversationText = context.messages
            .map((m) => `${m.role}: ${JSON.stringify(m.content)}`)
            .join("\n");
        const { text: summary } = await generateText({
            model,
            system: prompt,
            messages: [{ role: "user", content: conversationText }],
            abortSignal: context.signal,
        });
        const summaryMessages = [
            {
                role: "user",
                content: `[Previous conversation summary]\n\n${summary}\n\n[The conversation continues from here]`,
            },
        ];
        return {
            messages: summaryMessages,
            summary,
            messagesRemoved: context.messages.length - 1,
            tokensPruned: this.estimateTokens(context.messages) - this.estimateTokens(summaryMessages),
        };
    }
}
function estimateMessageTokens(msg, estimator) {
    return estimator([msg]);
}
function pruneToolResults(messages, protectedTokens, minSavings, estimateTokens) {
    const result = structuredClone(messages);
    let accumulated = 0;
    let boundary = result.length;
    // Find protection boundary (walk backward)
    for (let i = result.length - 1; i >= 0; i--) {
        accumulated += estimateMessageTokens(result[i], estimateTokens);
        if (accumulated >= protectedTokens) {
            boundary = i;
            break;
        }
    }
    // Prune tool results before boundary
    let tokensSaved = 0;
    let modified = 0;
    for (let i = 0; i < boundary; i++) {
        if (result[i].role === "tool") {
            const before = estimateMessageTokens(result[i], estimateTokens);
            const content = result[i].content;
            if (Array.isArray(content)) {
                result[i].content = content.map((part) => part.type === "tool-result" ? { ...part, result: "[pruned]" } : part);
            }
            const after = estimateMessageTokens(result[i], estimateTokens);
            tokensSaved += before - after;
            modified++;
        }
    }
    if (tokensSaved < minSavings) {
        return { messages, tokensSaved: 0, messagesModified: 0 };
    }
    return { messages: result, tokensSaved, messagesModified: modified };
}
const defaultEstimateTokens = _defaultEstimateTokens;
// ── Shared helpers (re-imported from utils.ts) ─────────────────────
const isRetryableError = _isRetryableError;
const getRetryDelay = _getRetryDelay;
const sleep = _sleep;
const addUsage = _addUsage;
// ── Session ─────────────────────────────────────────────────────────
export class Session {
    agent;
    sessionId;
    /** The conversation messages. Directly readable and replaceable. */
    messages = [];
    _turns = 0;
    _totalUsage = {
        inputTokens: 0,
        outputTokens: 0,
        totalTokens: 0,
    };
    contextWindow;
    reservedTokens;
    autoCompact;
    shouldCompactFn;
    compactionStrategy;
    retryConfig;
    hooks;
    sessionStore;
    lastInputTokens = 0;
    constructor(options) {
        this.agent = options.agent;
        this.sessionId = options.sessionId ?? crypto.randomUUID();
        this.contextWindow = options.contextWindow;
        this.reservedTokens =
            options.reservedTokens ?? Math.min(20_000, options.agent.maxTokens ?? 20_000);
        this.autoCompact = options.autoCompact ?? options.contextWindow !== undefined;
        this.shouldCompactFn = options.shouldCompact;
        this.compactionStrategy =
            options.compactionStrategy ?? new DefaultCompactionStrategy();
        this.retryConfig = { ...DEFAULT_RETRY, ...options.retry };
        this.hooks = options.hooks ?? {};
        this.sessionStore = options.sessionStore;
    }
    get turns() {
        return this._turns;
    }
    get totalUsage() {
        return this._totalUsage;
    }
    /** Load session from store. */
    async load() {
        if (!this.sessionStore)
            return false;
        const loaded = await this.sessionStore.load(this.sessionId);
        if (loaded) {
            this.messages = loaded;
            return true;
        }
        return false;
    }
    /** Save session to store. */
    async save() {
        if (!this.sessionStore)
            return;
        await this.sessionStore.save(this.sessionId, this.messages);
    }
    /** Send a message. Handles compaction + retry automatically. */
    async *send(input, options) {
        this._turns++;
        const turnNumber = this._turns;
        yield { type: "turn.start", turnNumber };
        // Auto-compaction check
        if (this.autoCompact && this.contextWindow && this.shouldCompactCheck()) {
            yield* this.compact({ signal: options?.signal });
        }
        // Hook: onBeforeSend
        let effectiveMessages = this.messages;
        if (this.hooks.onBeforeSend) {
            effectiveMessages = await this.hooks.onBeforeSend([...this.messages]);
        }
        const snapshot = [...this.messages];
        const { maxRetries } = this.retryConfig;
        const isRetryable = this.retryConfig.isRetryable ?? isRetryableError;
        let turnUsage = {
            inputTokens: undefined,
            outputTokens: undefined,
            totalTokens: undefined,
        };
        for (let attempt = 0; attempt <= maxRetries; attempt++) {
            let hasYieldedContent = false;
            let shouldRetry = false;
            try {
                for await (const event of this.agent.run(effectiveMessages, input, {
                    signal: options?.signal,
                })) {
                    if (event.type === "text.delta" || event.type === "tool.start") {
                        hasYieldedContent = true;
                    }
                    if (event.type === "step.done") {
                        this.lastInputTokens = event.usage.inputTokens ?? 0;
                    }
                    if (event.type === "error" &&
                        !hasYieldedContent &&
                        isRetryable(event.error) &&
                        attempt < maxRetries) {
                        const delayMs = getRetryDelay(attempt, this.retryConfig, event.error);
                        const suppressed = await this.hooks.onError?.(event.error, attempt);
                        yield {
                            type: "retry",
                            attempt,
                            maxRetries,
                            delayMs,
                            error: event.error,
                        };
                        this.messages = [...snapshot];
                        await sleep(delayMs, options?.signal);
                        shouldRetry = true;
                        break;
                    }
                    if (event.type === "done") {
                        this.messages = event.messages;
                        turnUsage = event.totalUsage;
                        yield event;
                        break;
                    }
                    yield event;
                }
            }
            catch (thrown) {
                const error = thrown instanceof Error ? thrown : new Error(String(thrown));
                if (!hasYieldedContent && isRetryable(error) && attempt < maxRetries) {
                    const delayMs = getRetryDelay(attempt, this.retryConfig, error);
                    await this.hooks.onError?.(error, attempt);
                    yield {
                        type: "retry",
                        attempt,
                        maxRetries,
                        delayMs,
                        error,
                    };
                    this.messages = [...snapshot];
                    await sleep(delayMs, options?.signal);
                    continue;
                }
                throw thrown;
            }
            if (shouldRetry)
                continue;
            break;
        }
        // Hook: onAfterResponse
        if (this.hooks.onAfterResponse) {
            await this.hooks.onAfterResponse({
                turnNumber,
                messages: this.messages,
                usage: turnUsage,
            });
        }
        // Persist
        if (this.sessionStore) {
            await this.sessionStore.save(this.sessionId, this.messages);
        }
        this._totalUsage = addUsage(this._totalUsage, turnUsage);
        yield { type: "turn.done", turnNumber, usage: turnUsage };
    }
    /** Manually trigger compaction. */
    async *compact(options) {
        const reason = this.shouldCompactCheck() ? "overflow" : "manual";
        const tokensBefore = defaultEstimateTokens(this.messages);
        yield { type: "compaction.start", reason, tokensBefore };
        // Get custom prompt from hook
        const context = {
            messages: this.messages,
            model: this.agent.model,
            systemPrompt: this.agent.systemPrompt,
            totalTokens: tokensBefore,
            targetTokens: this.contextWindow
                ? this.contextWindow - this.reservedTokens
                : tokensBefore / 2,
            signal: options?.signal,
        };
        if (this.hooks.onCompaction) {
            const customPrompt = await this.hooks.onCompaction(context);
            if (customPrompt) {
                context.compactionPrompt = customPrompt;
            }
        }
        const result = await this.compactionStrategy.compact(context);
        if (result.tokensPruned > 0) {
            yield {
                type: "compaction.pruned",
                tokensRemoved: result.tokensPruned,
                messagesRemoved: result.messagesRemoved,
            };
        }
        if (result.summary) {
            yield { type: "compaction.summary", summary: result.summary };
        }
        this.messages = result.messages;
        const tokensAfter = defaultEstimateTokens(this.messages);
        yield { type: "compaction.done", tokensBefore, tokensAfter };
    }
    /**
     * Convert a session turn into an AI SDK 5 UIMessage stream.
     * Returns a ReadableStream of typed UIMessageChunks that can be
     * passed to `createUIMessageStreamResponse()` or consumed directly.
     */
    toUIMessageStream(input, options) {
        return sessionEventsToUIStream(this.send(input, options), options);
    }
    /**
     * Convert a session turn into an HTTP Response with SSE encoding.
     * Ready to return from any HTTP handler (Next.js, Express, etc.).
     */
    toResponse(input, init) {
        const { signal, ...responseInit } = init ?? {};
        return createUIMessageStreamResponse({
            stream: this.toUIMessageStream(input, { signal }),
            ...responseInit,
        });
    }
    shouldCompactCheck() {
        if (!this.contextWindow)
            return false;
        const info = {
            lastInputTokens: this.lastInputTokens,
            contextWindow: this.contextWindow,
            reservedTokens: this.reservedTokens,
            messages: this.messages,
            turnNumber: this._turns,
        };
        if (this.shouldCompactFn) {
            return this.shouldCompactFn(info);
        }
        return this.lastInputTokens >= this.contextWindow - this.reservedTokens;
    }
}
