// @openharness/core/dist/todos.js
export const TODO_STATUSES = ["pending", "in_progress", "completed", "cancelled"];
export const TODO_PRIORITIES = ["high", "medium", "low"];
export class InMemoryTodoStore {
    entries = new Map();
    async load(sessionId) {
        const todos = this.entries.get(sessionId);
        return todos ? structuredClone(todos) : undefined;
    }
    async save(sessionId, todos) {
        this.entries.set(sessionId, structuredClone(todos));
    }
    async delete(sessionId) {
        this.entries.delete(sessionId);
    }
}
export function isTodoStatus(value) {
    return typeof value === "string" && TODO_STATUSES.includes(value);
}
export function isTodoPriority(value) {
    return typeof value === "string" && TODO_PRIORITIES.includes(value);
}
export function isTodoItem(value) {
    if (typeof value !== "object" || value === null)
        return false;
    const item = value;
    return ((item.id === undefined || typeof item.id === "string") &&
        typeof item.content === "string" &&
        isTodoStatus(item.status) &&
        isTodoPriority(item.priority));
}
export function parseTodoUpdate(output) {
    if (typeof output !== "object" || output === null)
        return undefined;
    const data = output;
    if (!Array.isArray(data.todos))
        return undefined;
    if (!data.todos.every(isTodoItem))
        return undefined;
    return {
        sessionId: typeof data.sessionId === "string" ? data.sessionId : "default",
        todos: structuredClone(data.todos),
    };
}
