// @openharness/core/dist/utils.js
export function isRetryableError(error) {
    const msg = error.message.toLowerCase();
    return (/429|500|502|503|504|529/.test(msg) ||
        msg.includes("rate limit") ||
        msg.includes("timeout") ||
        msg.includes("econnreset") ||
        msg.includes("overloaded"));
}
export function getRetryDelay(attempt, config, error) {
    const retryAfter = error.headers?.["retry-after"];
    if (retryAfter)
        return Math.min(parseInt(retryAfter) * 1000, config.maxDelayMs);
    const base = config.initialDelayMs * config.backoffMultiplier ** attempt;
    const jitter = Math.random() * 0.3 * base;
    return Math.min(base + jitter, config.maxDelayMs);
}
export function sleep(ms, signal) {
    return new Promise((resolve, reject) => {
        if (signal?.aborted) {
            reject(signal.reason);
            return;
        }
        const timer = setTimeout(resolve, ms);
        signal?.addEventListener("abort", () => {
            clearTimeout(timer);
            reject(signal.reason);
        }, { once: true });
    });
}
export function addUsage(a, b) {
    return {
        inputTokens: (a.inputTokens ?? 0) + (b.inputTokens ?? 0),
        outputTokens: (a.outputTokens ?? 0) + (b.outputTokens ?? 0),
        totalTokens: (a.totalTokens ?? 0) + (b.totalTokens ?? 0),
    };
}
export function defaultEstimateTokens(messages) {
    return Math.ceil(JSON.stringify(messages).length / 4);
}
