// @openharness/react/dist/provider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useState, useCallback, useRef } from "react";
import { OHContext, initialSessionState, initialTodoState, initialSandboxState, } from "./context.js";
/**
 * Provides shared OpenHarness state (subagent, session, sandbox) to all
 * child hooks. Place this at the root of your chat UI.
 *
 * The provider does NOT render any UI — it only manages state.
 */
export function OpenHarnessProvider({ children }) {
    const [subagents, setSubagents] = useState(() => new Map());
    const [sessionState, setSessionState] = useState(initialSessionState);
    const [todoState, setTodoState] = useState(initialTodoState);
    const [sandboxState, setSandboxState] = useState(initialSandboxState);
    // Use a ref to avoid re-creating dispatch on every render
    const subagentsRef = useRef(subagents);
    subagentsRef.current = subagents;
    const dispatch = useCallback((part) => {
        const data = part.data;
        switch (part.type) {
            // ── Subagent events ──────────────────────────────────────
            case "data-oh:subagent.start": {
                const path = data?.path ?? [data?.agentName ?? "unknown"];
                const key = path.join("/");
                const info = {
                    name: path[path.length - 1],
                    task: data?.task ?? "",
                    status: "running",
                    sessionId: data?.sessionId,
                    path,
                    startedAt: Date.now(),
                };
                setSubagents((prev) => {
                    const next = new Map(prev);
                    next.set(key, info);
                    return next;
                });
                break;
            }
            case "data-oh:subagent.done": {
                const path = data?.path ?? [data?.agentName ?? "unknown"];
                const key = path.join("/");
                setSubagents((prev) => {
                    const existing = prev.get(key);
                    if (!existing)
                        return prev;
                    const next = new Map(prev);
                    next.set(key, {
                        ...existing,
                        status: "done",
                        sessionId: data?.sessionId ?? existing.sessionId,
                        durationMs: data?.durationMs ?? Date.now() - existing.startedAt,
                    });
                    return next;
                });
                break;
            }
            case "data-oh:subagent.error": {
                const path = data?.path ?? [data?.agentName ?? "unknown"];
                const key = path.join("/");
                setSubagents((prev) => {
                    const existing = prev.get(key);
                    if (!existing)
                        return prev;
                    const next = new Map(prev);
                    next.set(key, {
                        ...existing,
                        status: "error",
                        sessionId: data?.sessionId ?? existing.sessionId,
                        error: data?.error ?? "Unknown error",
                    });
                    return next;
                });
                break;
            }
            // ── Session lifecycle events ─────────────────────────────
            case "data-oh:turn.start":
                setSessionState((prev) => ({
                    ...prev,
                    currentTurn: data?.turnIndex ?? prev.currentTurn + 1,
                }));
                break;
            case "data-oh:turn.done":
                // Turn completed — no special state change needed
                break;
            case "data-oh:session.compacting":
            case "data-oh:compaction.start":
                setSessionState((prev) => ({ ...prev, isCompacting: true }));
                break;
            case "data-oh:compaction.done":
                setSessionState((prev) => ({
                    ...prev,
                    isCompacting: false,
                    lastCompactionAt: new Date(),
                    messagesRemovedByCompaction: prev.messagesRemovedByCompaction +
                        (data?.messagesRemoved ?? 0),
                }));
                break;
            case "data-oh:retry":
                setSessionState((prev) => ({
                    ...prev,
                    isRetrying: true,
                    retryAttempt: data?.attempt ?? prev.retryAttempt + 1,
                    retryReason: data?.reason ?? null,
                }));
                break;
            // ── Todo events ─────────────────────────────────────────
            case "data-oh:todo.updated":
                setTodoState({
                    sessionId: typeof data?.sessionId === "string" ? data.sessionId : undefined,
                    todos: Array.isArray(data?.todos) ? data.todos : [],
                    updatedAt: new Date(),
                });
                break;
            // ── Sandbox events (future) ──────────────────────────────
            case "data-oh:sandbox.provisioning":
                setSandboxState((prev) => ({
                    ...prev,
                    isProvisioning: true,
                    provisioningMessage: data?.message ?? "Provisioning...",
                }));
                break;
            case "data-oh:sandbox.ready":
                setSandboxState((prev) => ({
                    ...prev,
                    isProvisioning: false,
                    isWarm: true,
                    provisionedAt: new Date(),
                    provisioningMessage: null,
                }));
                break;
        }
    }, []);
    const value = {
        subagents,
        sessionState,
        todoState,
        sandboxState,
        dispatch,
    };
    return _jsx(OHContext.Provider, { value: value, children: children });
}
