// @openharness/core/dist/types/stream-parts.js
// ── Type guards ─────────────────────────────────────────────────────
export const isOHDataPart = (part) => part.type.startsWith("data-oh:");
export const isSubagentEvent = (part) => part.type.startsWith("data-oh:subagent.");
export const isCompactionEvent = (part) => part.type.startsWith("data-oh:compaction.");
export const isRetryEvent = (part) => part.type === "data-oh:retry";
export const isTodoEvent = (part) => part.type === "data-oh:todo.updated";
export const isSessionLifecycleEvent = (part) => part.type.startsWith("data-oh:turn.") || part.type === "data-oh:session.compacting";
// ── SSE stream formatters ───────────────────────────────────────────
//
// These produce SSE-encoded strings matching the AI SDK 5 data stream
// protocol. Each returns a complete SSE event line (`data: ...\n\n`).
//
// Primary use: building raw ReadableStreams without `createUIMessageStream`.
// For the standard path, `toUIMessageStream` uses the AI SDK's writer
// and only needs the chunk objects (not SSE strings).
export function formatSSE(data) {
    return `data: ${JSON.stringify(data)}\n\n`;
}
export function formatTextDelta(id, delta) {
    return formatSSE({ type: "text-delta", id, delta });
}
export function formatReasoningDelta(id, delta) {
    return formatSSE({ type: "reasoning-delta", id, delta });
}
export function formatDataPart(name, data) {
    return formatSSE({ type: `data-${name}`, data });
}
export function formatToolStart(id, toolName) {
    return formatSSE({ type: "tool-input-start", id, toolName });
}
export function formatToolResult(id, output) {
    return formatSSE({ type: "tool-output-available", id, output });
}
export function formatToolError(id, error) {
    return formatSSE({ type: "tool-output-error", id, error });
}
export function formatStepStart(messageId) {
    return formatSSE({ type: "start-step", ...(messageId ? { messageId } : {}) });
}
export function formatStepFinish(finishReason, usage, messageId) {
    return formatSSE({
        type: "finish-step",
        finishReason,
        ...(usage ? { usage } : {}),
        ...(messageId ? { messageId } : {}),
    });
}
export function formatFinishMessage(finishReason, usage) {
    return formatSSE({ type: "finish", finishReason, ...(usage ? { usage } : {}) });
}
export function formatDone() {
    return "data: [DONE]\n\n";
}
