// @openharness/react/dist/hooks/use-open-harness.js
import { useEffect, useRef } from "react";
import { useChat } from "@ai-sdk/react";
import { useOHContext } from "../context.js";
import { createOHTransport } from "../transport.js";
/**
 * Primary hook for OpenHarness chat. Wraps AI SDK 5's `useChat` with:
 * - `OHUIMessage` as the message type
 * - Transport pre-wired to your endpoint
 * - Data parts routed into the OpenHarnessProvider's event dispatcher
 *
 * Must be used within an `<OpenHarnessProvider>`.
 */
export function useOpenHarness(config) {
    const { dispatch } = useOHContext();
    const chat = useChat({
        // Only pass `id` when explicitly provided. Passing `undefined` causes
        // AI SDK's useChat to recreate the Chat instance on every render because
        // the auto-generated id never matches the `undefined` in options.
        ...(config.id != null && { id: config.id }),
        messages: config.messages,
        resume: config.resume,
        transport: createOHTransport(config.endpoint, config),
        onData: (part) => dispatch(part),
        onFinish: config.onFinish,
    });
    const hydratedChatIdRef = useRef(null);
    useEffect(() => {
        if (!config.messages?.length)
            return;
        if (hydratedChatIdRef.current === chat.id)
            return;
        if (chat.messages.length > 0)
            return;
        chat.setMessages(config.messages);
        hydratedChatIdRef.current = chat.id;
    }, [chat.id, chat.messages.length, chat.setMessages, config.messages]);
    return chat;
}
